File: Basic data structure/LinkedList.py
#implement doubly linked list
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None
class LinkedList:
    def __init__(self):
        self.head = None
    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        cur = self.head
        while cur.next:
            cur = cur.next
        cur.next = new_node
        new_node.prev = cur
    def append_before_last(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        cur =self.head
        prev=None
        while cur.next!=None:
            prev=cur
            cur=cur.next
        temp=prev.next
        prev.next=new_node
        new_node.prev=prev
        new_node.next=temp
    
    def return_list(self):
        cur = self.head
        lst=[]
        while cur:
            lst.append(cur.data)
            cur = cur.next
        print(lst)
        return lst
    def search(self, data):
        cur = self.head
        while cur:
            if cur.data == data:
                return True
            cur = cur.next
        return False
    def pop(self):
        if self.head is None:
            print("List is empty")
            return
        cur = self.head
        while cur.next:
            cur = cur.next
        if(cur.prev==None):
            self.head=None
        else: 
            cur.prev.next = None
        return cur.data
    def delete_first(self):
        if self.head is None:
            print("List is empty")
            return
        cur = self.head
        self.head = cur.next
        return cur.data
    def delete(self, data):
        node=self.search(data)
        if node==False:
            print("Node not found")
            return
        if node.prev==None:
            self.head=node.next
            return
        else:
            node.prev.next=node.next

        if node.next!=None:
            node.next.prev=node.prev

class Stack:
    def __init__(self):
        self.ll = LinkedList()
    def push(self, data):
        self.ll.append(data)
    def pop(self):
        return self.ll.pop()
    def return_list(self):
        return self.ll.return_list()

class Queue:
    def __init__(self):
        self.ll = LinkedList()
    def enqueue(self, data):
        self.ll.append(data)
    def dequeue(self):
        return self.ll.delete_first()
    def return_list(self):
        return self.ll.return_list()
#test LinkedList
llist = LinkedList()
llist.append("A")
llist.append("B")
llist.append_before_last("C")
assert llist.return_list() == ["A", "C", "B"]
#test stack
stack = Stack()
stack.push("A")
stack.push("B")
assert stack.pop() == "B"
assert stack.pop() == "A"
#test queue
queue = Queue()
queue.enqueue("A")
queue.enqueue("B")
assert queue.dequeue() == "A"
assert queue.dequeue() == "B"

            
        

File: Basic data structure/java/linkedlist.java
class Main {
    public static void main(String[] args) {
        SinglyLinkedList list = new SinglyLinkedList();
        list.insert(10);
        list.insert(20);
        list.insert(30);
        list.insertAfter(40);
        list.print();
        list.delete(20);
        list.print();
        DoubleLinkedList list2 = new DoubleLinkedList();
        list2.insert(10);
        list2.insert(20);
        list2.insert(30);
        list2.insertAfter(40);
        list2.print();
        list2.delete(20);
        list2.print();

    }
}

class Node {
    int data;
    Node next;
    Node prev;

    public Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }

}

// singly linked list
public class SinglyLinkedList {
    Node head;

    public SinglyLinkedList() {
        this.head = null;
    }

    public Node search(int n) {
        Node x = head;
        while (x != null && x.data != n) {
            x = x.next;
        }
        if (x == null) {
            return null;
        }
        return x;
    }

    public void insert(int n) {
        Node x = new Node(n);
        x.next = head;
        head = x;
    }

    // insert at the end
    public void insertAfter(int n) {
        Node x = new Node(n);
        if (head == null) {
            head = x;
            return;
        }
        Node y = head;
        while (y.next != null) {
            y = y.next;
        }
        y.next = x;
    }

    public void delete(int n) {
        if (head == null) {
            return;
        }
        if (head.data == n) {
            head = head.next;
            return;
        }
        Node x = head;
        while (x.next != null && x.next.data != n) {
            x = x.next;
        }
        if (x.next == null) {
            return;
        }
        x.next = x.next.next;
    }

    public void print() {
        Node x = head;
        while (x != null) {
            System.out.printf("%d ", x.data);
            x = x.next;
        }
        System.out.println();
    }

}

class DoubleLinkedList {
    Node head;

    public DoubleLinkedList() {
        this.head = null;
    }

    public Node search(int n) {
        Node x = head;
        while (x != null && x.data != n) {
            x = x.next;
        }
        if (x == null) {
            return null;
        }
        return x;
    }

    public void insert(int n) {
        Node x = new Node(n);
        x.next = head;
        if (head != null) {
            head.prev = x;
        }
        head = x;
    }

    public void insertAfter(int n) {
        Node x = new Node(n);
        if (head == null) {
            head = x;
            return;
        }
        Node y = head;
        while (y.next != null) {
            y = y.next;
        }
        y.next = x;
        x.prev = y;
    }

    public void delete(int n) {
        if (head == null) {
            return;
        }
        if (head.data == n) {
            head = head.next;
            return;
        }
        Node x = head;
        while (x.next != null && x.next.data != n) {
            x = x.next;
        }
        if (x.next == null) {
            return;
        }
        x.next = x.next.next;
        if (x.next != null) {
            x.next.prev = x;
        }
    }

    public void print() {
        Node x = head;
        while (x != null) {
            System.out.printf("%d ", x.data);
            x = x.next;
        }
        System.out.println();
    }
}